[COT] Log errors during the posts to orders table migrations

Konamiman opened the pull request:
### All Submissions:

-   [ ] Have you followed the [WooCommerce Contributing guideline](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md)?
-   [ ] Does your code follow the [WordPress' coding standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/)?
-   [ ] Have you checked to ensure there aren't other open [Pull Requests](../../pulls) for the same update/change?

### Changes proposed in this Pull Request:

Implement logging for errors that happen during the execution of `PostsToOrdersMigrationController::migrate_orders`. The following types of errors are logged:

1. Unhandled exceptions.
2. Database errors caused by `$wpdb->query` and `$wpdb->get_results`.
3. Not enough rows processed after an insert or update operation.
4. Validation errors on the data to be migrated.

Errors are logged using the standard WooCommerce logger, using `posts-to-orders-migration` as the source name.

Closes #32664.

### How to test the changes in this Pull Request:

First, enable the feature by adding the following code snippet somewhere, for example at the end of `woocommerce.php`:

```php
wc_get_container()
   ->get('Automattic\WooCommerce\Internal\DataStores\Orders\CustomOrdersTableController')
   ->show_feature();
```

Then force errors of each type, then go to WooCommerce - Status - Logs and verify that the errors are properly logged.

One way to force errors is to to temporarily modify the code as follows (don't try more than one at the same time tas it would get confusing):

1. In `TableMigrator::process_migration_batch_for_ids`, right before the first statement of the `try` block, add this: `throw new \Exception("Ouch!!");`
2. In `TableMigrator::db_query`, right before the `return` statement, add this: `$this->add_error('Gasp!!');` (also in `db_get_results`)
3. In `TableMigrator::maybe_add_insert_or_update_mismatch_error`, right after `$count` is calculated increase it by one.
4. In `MetaToCustomTableMigrator::validate_data`, add this right before the first line of the method: `return new \WP_Error('Argh!!');`

Feel free to use other (probably more proper) ways to force errors. For example you could temporarily remove or rename the `wp_wc_orders` table.

### Other information:

-   [ ] Have you added an explanation of what your changes do and why you'd like us to include them?
-   [ ] Have you written new tests for your changes, as applicable?
-   [ ] Have you successfully run tests with your changes locally?
-   [ ] Have you created a changelog file for each project being changed, ie `pnpm nx changelog <project>`?

### FOR PR REVIEWER ONLY:

-   [ ] I have reviewed that everything is sanitized/escaped appropriately for any SQL or XSS injection possibilities. I made sure Linting is not ignored or disabled.

Konamiman committed:
Adjust visibility of methods in the orders table migrator classes.

Many methods were public/protected when they could be protected/private
instead. All methods are adjusted to the minimum visibility required
for the current implementation to work.

Also minor adjustments for phpcs, such as "ignore" comments.

Konamiman committed:
Fix issues in error handling in the MetaToCustomTableMigrator class

There were some inconsistencies, such as trying to treat an error
array as a string, or accessing to a non-existing column from a
dataset retrieved from the db on error.

Konamiman committed:
Log all db errors and exceptions in the posts to orders migrators.

To achieve this:

- A new class is added, TableMigrator, containing basic in-memory
  error storage methods, and auxiliary db_query and db_get_result
  methods.
- MetaToCustomTableMigrator and MetaToMetaTableMigrator classes
  now extend TableMigrator and use these db_* methods.
- process_migration_batch_for_ids is now a method in TableMigrator,
  it just calls the abstract process_migration_batch_for_ids_core,
  catches exceptions, and returns an array of errors.
- PostsToOrdersMigrationController now logs all the errors returned
  by process_migration_batch_for_ids in all the migrator classes.
- New utility method: ArrayUtil::to_ranges_string

Konamiman committed:
Fix: MetaToMetaTableMigrator attempting to migrate empty datasets

That caused an invalid SQL query to throw an error.

Konamiman committed:
Merge branch 'trunk' into add/logging_for_posts_to_cot_migration

botwoo commented:
:bar_chart: Test reports for 4b0d5550edb7a9f005395b7f0ccc608efcf0715f `Merge branch 'trunk' into add/logging_for_posts_to_cot_migration` have been published and are accessible through the following links:

- [REST API Test Report](https://woocommerce.github.io/woocommerce-test-reports/pr/33034/api/)
- E2E Test Report (TBC)

Visit the [WooCommerce Test Reports](https://woocommerce.github.io/woocommerce-test-reports) homepage to view all published reports. See the [FAQs](https://github.com/woocommerce/woocommerce-test-reports/wiki/FAQs) page if you're having problems accessing them.

vedanshujain commented on plugins/woocommerce/src/Database/Migrations/CustomOrderTable/MetaToCustomTableMigrator.php:
```diff
@@ -224,21 +216,23 @@ protected function generate_column_clauses( array $columns_schema, array $batch
+	 * Migrate a batch of orders from the posts table to the corresponding table.
 	 *
-	 * @param array $entity_ids List of entity IDs to perform migrations for.
+	 * @param array $entity_ids Ids of orders ro migrate.
```
typo  `ro` -> `to`
```suggestion
	 * @param array $entity_ids Ids of entities to migrate.
```

vedanshujain commented on plugins/woocommerce/src/Database/Migrations/CustomOrderTable/MetaToCustomTableMigrator.php:
```diff
@@ -224,21 +216,23 @@ protected function generate_column_clauses( array $columns_schema, array $batch
 
 	/**
-	 * Process next migration batch, uses option `wc_cot_migration` to checkpoints of what have been processed so far.
+	 * Migrate a batch of orders from the posts table to the corresponding table.
```
suggestion to use `entities` instead of orders since this class (along with MetaToMetaTableMigrator) are generic, and they can migrate any entity (product, coupon, booking, etc) from the posts table and not just orders. On that note, we should also move them out of the `CustomOrderTable` folder.

vedanshujain commented on plugins/woocommerce/src/Utilities/ArrayUtil.php:
```diff
@@ -67,5 +67,48 @@ public static function is_truthy( array $array, string $key ) {
+	 *
+	 * @return string
+	 */
+	public static function to_ranges_string( array $items, string $item_separator = ', ', string $range_separator = '-', bool $sort = true ): string {
```
nice

vedanshujain commented on plugins/woocommerce/src/Database/Migrations/CustomOrderTable/PostsToOrdersMigrationController.php:
```diff
@@ -18,86 +19,102 @@ class PostsToOrdersMigrationController {
+			$meta_table_migrator,
+		);
+
+		$this->error_logger = wc_get_logger();
```
we can remove the `MigrationErrorLogger` if we are not going to use it

vedanshujain reviewed (commented):
Looks good to me. I have added some minor suggestions. Other than those, I was wondering if we can add some tests for the logger as well? For example, we can delete tables and execute the migration to see what will come in logs.

Konamiman committed:
Remove the MigrationErrorLogger class as it's not used.

Konamiman committed:
Change references to "orders" to "entities" in generic table migrators

Konamiman committed:
Move the base table migration classes outside CustomOrderTable directory

These classes are generic and don't depend on the custom orders table
functionality, so they are moved from src/Database/Migrations/CustomOrderTable
to just src/Database/Migrations.

Konamiman commented:
Changelog file added in https://github.com/woocommerce/woocommerce/pull/33136
